Document passable records (#506)
Agoric/documentation: main/glossary/README.md

## Comparable

A *comparable* is a [passable](#passable) whose leaves contain no promises. Two 
comparables can be synchronously compared for structural equivalence. If either passable

> object contains Promises, equality is indeterminable. If both are fulfilled down

Reviewer: I think this can be made a bit clearer. How about 
"A *comparable* is a [passable](#passable) which does not contain any Promises, It can be locally and synchronously compared to another object for equality, including structural equivalence. But if either passable contains Promises, equality cannot be determined. If both objects are fulfilled down to Presences and local state, then either they're the same all the way down, or they represent different objects."

escrowed. One reason this is possible is if a [proposal](#proposal) doesn't match what the contract expects to do, it
can immediately cause the [seat](#seat) to exit, getting back the amount it offered.

## Passable

A *passable* is something that can be marshalled.

There are three kinds of passables:

>    * Remotables, objects with methods that can be called remotely using `E()`, and their remote Presence.

Reviewer: Link E() to its glossary entry #e.

   * Pass-by-copy data, such as numbers or hardened records.

>    * Promises for passables.

Reviewer: This is a bit circular. How about "Promises for either remotables or pass-by-copy data." to be clearer?

Reviewer 2: Saying "pass-by-copy data" would exclude remotables. But promises can be fulfilled to remotables, so that would not be correct. 

The truth is exactly as circular as it sounds. A promise can be fulfilled to any passable. To be fully accurate, a promise can be fulfilled to any non-promise passable.

> For more information, see the [Remotable and passable objects documentation](/guides/js-programming/far.md#remotable-and-passable-objects).

Reviewer 2: Link seems broken, at least when I try to follow it in github preview. In any case, I see the intended target below. So if this is the correct way to link to it, no change suggested.

Reviewer: Links get changed when Vuepress processes the Markdown into HTML for display, at the very least from .md to .html.

Reviewer: Added a "Remotable" entry.

A transfer of [amounts](#amounts) between [seats](#seat) within Zoe; i.e. a change in their [allocations](#allocation). When a seat exits, it gets its
current allocation as a [payout](#payout). 

## Record and Tuple

> Records and tuples are immutable objects which can be passed by copy. They only contain values which can be passed by copy such as numbers, strings, or other records and tuples. 

Reviewer 3: Hmm, they can contain presences/remotables (what's the right term again?) as values, so we should probably mention that. Otherwise I think this description is misleading.

Author: Oh I didn't realize that remotables could be nested inside pass by copy structures! And it is definitely not obvious from the documentation.
https://agoric.com/documentation/guides/js-programming/far.html

A transfer of [amounts](#amounts) between [seats](#seat) within Zoe; i.e. a change in their [allocations](#allocation). When a seat exits, it gets its
current allocation as a [payout](#payout). 

## Record and Tuple

Records and tuples are immutable objects which can be passed by copy. They only contain values which are [passables](#passable). 

> For more details, see [Passable objects documentation](/guides/js-programming/far.md#rules-for-creating-passable-objects).

Reviewer: Suggest adding "see the full..."

Records and tuples are immutable objects which can be passed by copy. They only contain values which are [passables](#passable). 
For more details, see [Passable objects documentation](/guides/js-programming/far.md#rules-for-creating-passable-objects).

## Remotable

> Remotables are objects intended to be used from other vats. Remote messages sent to remotables must only contain [passable](#passable)
> arguments and return passable results. 
> 
> All of a remotable's property values must be functions, and cannot be accessors. Its methods are called by wrapping 
> the remotable object with [`E`](#e), such as `E(issuer).getBrand()`. As with all calls to `E()`, it results in a Promise.
> 
> Every object returned from a smart contract, such a publicFacet or creatorFacet, must be passable. All objects used in 
> your contract's external API must be passable.

Author: I feel like there is a bit of repetition here, and slight confusion between remotable and presence.

Wondering if we could structure it the following way:
```suggestion
Remotables are objects containing only methods, intended to be used from other vats. The methods of remotables accept only [passable](#passable) arguments and return passable results. 

A remotable is used in other vats through its proxy called a [presence](#presence). A remotable's methods are called by wrapping the presence with [`E`](#e), such as `E(issuer).getBrand()`.
```

The last line feels more appropriate for the passable section.

@tyg, what do you think? Happy to include this proposed change when working on the other feedback.

Reviewer 2: Since "Proxy" has a specific meaning in JS, and presences are not JS proxies, please avoid describing them using "proxy" terminology. "Representative" is good. (Our own internal use of "representative" for a different concept should not stop us. We can change our unexplained internal terminology if needed.)

Author: there are apparently other uses of the term proxy in relation to presence. I'll double check those.

Reviewer 2: Actually, both the old text and the new text are not quite right on the Remotable/presence terminology. See https://github.com/Agoric/agoric-sdk/pull/2909/files . Warning: Pedantic clarification ahead. Almost certainly too much detail to actually include here. But it should constrain the text we do include here to be correct.

The thing with methods is a PrimaryRemotable. The remote presences are Remotes. Both PrimaryRemotables and Remotes are Remotables. At this level of abstraction, we talk about Remotes, but all Remotes are presences. Not all presences are Remotes though. It is the eventual-send level of abstraction that defines presences. Captp uses presences only to implement Remotes. However, other code is free to use eventual-send to create presences that are not Remotes.

Reviewer 2: Attn @michaelfig , with whom we worked out the precise terminology.
